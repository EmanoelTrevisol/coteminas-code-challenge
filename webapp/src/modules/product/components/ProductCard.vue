<template>
  <div class="product-card">
    <div class="images">
      <div
        class="image"
        v-for="(image, key) in product.images"
        :key="`image-${key}-${product._id}`"
      >
        <img :src="image" :alt="`Imagem ${key} do produto ${product.title}`" />
      </div>
    </div>
    <div class="info">
      <div class="title">
        <div class="product-name">
          {{ product.title }}
        </div>
        <small class="product-type-size">
          {{ product.type }} | {{ product.size }}
        </small>
      </div>
      <p class="price">
        <small v-if="product.offerPrice" class="original">
          <span class="currency">
            {{ product.price }}
          </span>
          por
        </small>
        {{ product.offerPrice || product.price }}
      </p>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    product: {
      type: Object,
      default: () => ({}),
    },
  },
};
</script>

<style lang="stylus" scoped>
.product-card {
  padding: 5px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  border: 1px solid $light;

  .images {
    padding: 5px;
    display: flex;
    justify-content: space-evenly;
    align-items: center;

    .image {
      width: 96px;
      height: 96px;
      margin: 5px;
    }
  }

  .info {
    display: flex;
    justify-content: space-between;
    align-items: center;
    width: 100%;
    padding: 10px 20px;

    .title {
      .product-name {
        color: $dark;
        font-size: 1.1em;
      }

      .product-type-size {
        color: $medium;
      }
    }

    .price {
      font-size: 1.1em;
      text-align: end;
      .original {
        color: $medium;
        .currency {
          text-decoration: line-through;
        }
      }

      .main {
        color: $medium;
      }
    }
  }
}
</style>
