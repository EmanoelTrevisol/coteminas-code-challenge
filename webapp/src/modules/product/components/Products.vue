<template>
  <div class="products">
    <product-card
      v-for="product in products"
      :key="product._id"
      :product="product"
    />
  </div>
</template>

<script>
import ProductCard from './ProductCard';

export default {
  components: {
    ProductCard,
  },
  props: {
    products: {
      type: Array,
      default: () => [],
    },
  },
};
</script>
