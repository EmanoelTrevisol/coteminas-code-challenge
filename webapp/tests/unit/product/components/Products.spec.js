import Products from '@/modules/product/components/Products';
import ProductCard from '@/modules/product/components/ProductCard';
import { shallowMount } from '@vue/test-utils';
import mockedProducts from '@/modules/product/_mockedDBProducts/productsArray';

describe('Products.vue', () => {
  describe('renders Products.vue', () => {
    const build = (propsData) => {
      const wrapper = shallowMount(Products, {
        propsData: propsData || {
          products: [...mockedProducts],
        },
      });

      return wrapper;
    };

    it('should match snapshot', () => {
      const wrapper = build();
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render ProductCard', () => {
      const wrapper = build();

      const cards = wrapper.findAllComponents(ProductCard);

      expect(cards.length).toBe(mockedProducts.length);
    });

    it('should pass product prop correctly to ProductCard', () => {
      const baseProduct = mockedProducts[0];
      const wrapper = build({ products: [{ ...baseProduct }] });

      const cards = wrapper.findAllComponents(ProductCard);

      const card = cards.at(0);

      expect(cards.length).toBe(1);
      expect(card.vm.product).toStrictEqual({ ...baseProduct });
    });
  });
});
