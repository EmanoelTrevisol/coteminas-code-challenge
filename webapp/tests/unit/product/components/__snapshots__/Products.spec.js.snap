// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products.vue renders Products.vue should match snapshot 1`] = `
<div class="products">
  <product-card-stub product="[object Object]"></product-card-stub>
  <product-card-stub product="[object Object]"></product-card-stub>
  <product-card-stub product="[object Object]"></product-card-stub>
  <product-card-stub product="[object Object]"></product-card-stub>
  <product-card-stub product="[object Object]"></product-card-stub>
  <product-card-stub product="[object Object]"></product-card-stub>
  <product-card-stub product="[object Object]"></product-card-stub>
  <product-card-stub product="[object Object]"></product-card-stub>
  <product-card-stub product="[object Object]"></product-card-stub>
  <product-card-stub product="[object Object]"></product-card-stub>
  <product-card-stub product="[object Object]"></product-card-stub>
</div>
`;
